### ISSUE TITLE
Regression in DateTimeZone::getTransitions()

### ISSUE THREAD

**Issue by ssandler** [facebook/hhvm#8868]
**Describe the bug**
`DateTimeZone::getTransitions()` no longer returns all transitions, it returns only the first one.

**Standalone code, or other way to reproduce the problem**
```
var_dump(new DateTimeZone('America/Los_Angeles')->getTransitions(1420070400, 1577836800));
```

On HHVM 4.80 (and any version up to 4.91) this returns 11 records. Starting with 4.92 and including latest, this returns only the first record. As far as I can tell that's the case for all possible timezone and date inputs, only one record is ever returned. I bisected this using hhvm public docker images.

Looking at `git log HHVM-4.91.0..HHVM-4.92.0` I see https://github.com/facebook/hhvm/commit/21ee3f8e3fe2eb274f575c3817d6a95b823fe378 which is a possible candidate. I also noticed there are test cases in HHVM for this such as `hphp/test/slow/ext_datetime/date_timezone_transitions.php` which would fail if only one transition is returned (and those tests do fail for me on the public build, `var_dump($transitions[3]);` throws since there's only one entry). If those tests pass for FB I wonder if you have something internally that is making this work? A different version of tzdata perhaps?

**Expected behavior**

All timezone transitions are returned.

**Actual behavior**

Only one is returned.

**Environment**
 - Operating system
Ubuntu 20.04
 - Installation method
hhvm/hhvm on dockerhub
 - HHVM Version
```
HipHop VM 4.122.0 (rel) (non-lowptr)
Compiler: 1628699243_156194841
Repo schema: 10a34637ad661d98ba3344717656fcc76209c2f8
```

**Comment by fredemmott** [comment]
The nightly docker images also match that commit; if it's that, most likely possibilities are:
- incompatible tzdata and we need to handle more formats
- third-party/forks/timelib/ is too outdated and needs updating (seems most likely)
- potentially related to either of the above: we might be using the build-in tzdata instead of system tzdata. We should probably make this a hard error unless a specific CMake option is set.

This issue does not repro on FB servers. Key differences are the version of timelib, and that we package tzdata separately.

**Comment by fredemmott** [comment]
And some notes for investigating:
- PHP can be a bit ambiguous for debugging this: by default, it bundles tzdata (with upgrades via pecl), but *most* packages change it to use system tzdata
- The system tzdata can be inspected on linux and macos with ` zdump -v 'America/Los_Angeles' | grep '201[56789]'`

**Comment by ssandler** [comment]
Tested a local build of HHVM 4.102 reverting https://github.com/facebook/hhvm/commit/21ee3f8e3fe2eb274f575c3817d6a95b823fe378 and it did resolve the problem.

**Comment by fredemmott** [comment]
FWIW we do know that without that patch, some newer tzdata (late 2020+) can not be loaded at all. It's not a suitable approach for us to put into our packages, but if it works in your environments, seems reasonable to do locally

**Comment by ssandler** [comment]
So as noted in the above PR, updating timelib doesn't seem to resolve the problem. I confirmed in my build that timelib 2021.07 is built and linked, but I still only get one result for `getTransitions()`.

It's worth noting that even reverting https://github.com/facebook/hhvm/commit/21ee3f8e3fe2eb274f575c3817d6a95b823fe378, I'm able to load the 2021a tzdata package. You mentioned it's possible we might be loading (old) built-in tzdata. I confirmed we aren't by:
```
dpkg -l | grep tzdata
ii  tzdata                           2021a-0ubuntu0.18.04                  all          time zone and daylight-saving time data
```

```
hhvm -a
hphpd> $tz = timezone_open('Africa/Juba')
hphpd> =$tz->getOffset(new DateTime('now', $tz))
7200
```

The https://bugs.launchpad.net/ubuntu/+source/tzdata/+bug/1913482 release notes indicate that this timezone was updated from UTC+3 to UTC+2 in the 2021a release, so this makes me believe that hhvm is in fact loading the system tzdata, and that tzdata newer than late 2020 still works without https://github.com/facebook/hhvm/commit/21ee3f8e3fe2eb274f575c3817d6a95b823fe378, both on HHVM 4.80.6 and also on HHVM 4.102 if I revert that commit.

**Comment by ssandler** [comment]
So I believe I have this all figured out, however I am not an experience c/c++ programmer so definitely correct where I'm wrong.

1. The current timelib that HHVM has is too old for `bit64` to be fully populated. There are actually newer commits on the `hhvm-third-party` repo, but even origin/master of hhvm/third-party/forks is submoduled to target https://github.com/hhvm/hhvm-third-party/tree/bddd0f2f19acc6bda8eec9a29657e0e041d290c3 which doesn't have those commits. The code that would read the 64bit tzdata is skipped by the current version HHVM is using due to the patch to use system tzdata https://github.com/hhvm/hhvm-third-party/blob/bddd0f2f19acc6bda8eec9a29657e0e041d290c3/timelib/parse_tz.c#L885

That #ifdef puts all of this into an else block:
```
skip_64bit_preamble(&tzf, tmp);
read_64bit_header(&tzf, tmp);
skip_64bit_transistions(&tzf, tmp);
skip_64bit_types(&tzf, tmp);
skip_posix_string(&tzf, tmp);
```

2. I updated timelib and it didn't resolve this problem, but that was because I lost the changes to make hhvm use system tzdata. It was instead using tzdata compiled into timelib, which had really old transitions. That's why my code snippet from this issue didn't return >1 record still. When I changed the first arg to `0` I got plenty of results.
3. I figured that out and updated hhvm's `timezone.cpp` to load `/usr/share/zoneinfo` if present. timelib got this functionality 4 years ago in https://github.com/derickr/timelib/commit/1ccf47679b50340328845223dea99647ee94e9eb, so we don't need to patch it in anymore.

The updated PR linked [above](https://github.com/facebook/hhvm/pull/8870) now does resolve this for me - I built 4.102 with those changes and got the expected 11 transitions from the snippet in this issue.
